File: cpp/124_二叉树中的最大路径和_hard.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int ans = INT_MIN;
        oneSideMax(root, ans);
        return ans;
    }
    int oneSideMax(TreeNode *root, int &ans){
        if(root == nullptr)
            return 0;
        int left = max(0, oneSideMax(root->left, ans));
        int right = max(0, oneSideMax(root->right, ans));
        ans = max(ans, left+right+root->val);
        return max(left, right) + root->val;
    }
};

File: python/124_二叉树中的最大路径和_hard.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: TreeNode) -> int:
        def oneSideMax(root, ans):
            if root == None:
                return 0
            # oneSideMax(root.left, ans)[0]
            left = max(0, oneSideMax(root.left, ans))
            right = max(0, oneSideMax(root.right, ans))
            ans[0] = max(ans[0], left+right+root.val)
            return max(left, right) + root.val

        ans = [float('-inf')]
        oneSideMax(root, ans)
        return ans[0]